### PR TITLE
new core datastructure: CircularArray

### PR DESCRIPTION
**Pull request by privat**
`lib/core` is extended with CircularArray for an array with efficient push/pop/shift/unshift

Some number with bench_seq. I used valgrind with n=1000

push (Ir/call)
* list: 383
* unrolled list: 188
* circular array: 135
* array: 85

pop
* unroll: 180
* list: 79
* array: 66
* circ: 47

unshift
* array: 31,755 (because memmove)
* list: 385
* unroll: 322
* circ: 127

shift
* array: 36,168 (because memmove)
* unroll: 180
* list: 79
* circ: 50

insert:
* circ: 179,811 (because manual copy item by item)
* array: 27,282
* unroll: 12,235
* list: 11,500

### CODE REVIEW

File: lib/core/collection/circular_array.nit [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+#
+# A circular array offers efficient random access,
+# efficient manipulation for both ends of the structure (push, pop, ) and
+# automatic amortized growth.
+#
+# Therefore it can be used as is or as and efficient queue (FIFO/LIFO)
+class CircularArray[E]
+	super Sequence[E]
+
+	# The
```
Comment (R4PaSs) [inline lib/core/collection/circular_array.nit]: Why the isolated `The` ?

**Comment by R4PaSs** [comment]
Seems legit, :shipit:

**Comment by jpcaissy** [comment]
:+1:

**Comment by privat** [comment]
reroll: fix comments

**Comment by ablondin** [comment]
Good !

**Pull request merged**